assignment2/binarytree.py

```diff
@@ -0,0 +1,144 @@
+                               `None` by default, that means that any node in
+                               the tree could be a parent of a new one;
+            isLeft (optional): a bool, True if a new node should be a left child
+                               (if it possible) and False if it should be a
+                               right child (if it possible). `None` by default,
+                               that means it doesn't matter which child would be
+                               a new node. If `parent = None` then `isLeft`
+                               won't be taken into account.
+
+        Return:
+            a bool, True if the new key was successfully inserted in the tree
+            (even if conditions `parent` and `isLeft` are violated), and False
+            otherwise.
+        """
+        if self.root is None:
+            self.root = Node(key)
+            self.size += 1
+            return True
+        else:
+            if self.find(key):
+                return False  # duplicated keys are not allowed
+            if parent is None or not self.find(parent):
+                parent = isLeft = None
+            # check if it possible to insert a new node to this parent
+            sucsees = self._add(key, self.root, parent, isLeft)
+            if not sucsees:
+                # just insert it somewhere (always possible)
+                self._add(key, self.root)
+            self.size += 1
+            return True
+
+    def _add(self, key, cur, parent = None, isLeft = None):
+        """Internal iplementation of `insert(...)`.
+
+        Args:
+            key:               an integer, new key that should be inserted;
+            cur:               a Node, current node in the tree that could be a
+                               parent of a new one;
+            parent (optional): an integer, key of some node in the tree, that
+                               should be a parent of a new node, if it possible.
+                               `None` by default, that means that any node in
+                               the tree could be a parent of a new one;
+            isLeft (optional): a bool, True if a new node should be a left child
+                               (if it possible) and False if it should be a
+                               right child (if it possible). `None` by default,
+                               that means it doesn't matter which child would be
+                               a new node.
+
+        Return:
+            a bool, True if the new key was successfully adde to the current
+            node or some of its childs, and False otherwise
+        """
+        if cur is None:
+            return False
+        if parent is not None:
+            if cur.key != parent:
+                return (self._add(key, cur.left, parent, isLeft) or
+                        self._add(key, cur.right, parent, isLeft))
+            if cur.right is not None and cur.left is not None:
```

Comment:
@flerdacodeu I modified it a little bit, please check it. If `isLeft` is specified but such doesn't abailable than I raised an error. Otherwise, I insert new node as you said.